### ISSUE TITLE
Getting the values of hidden layer nodes

### ISSUE THREAD

**Issue by krayush07** [tflearn/tflearn#194]
Here's a code for XOR in tflearn. I wish to get the values of penultimate hidden layers nodes (and not the weights). I am able to get the weights and biases but unable to compute/get the values of hidden layer nodes, how can I get that? More specifically, I wish to obtain the values of layer2 (given in the code) nodes for each of the four predictions as given below.

```python
import tensorflow as tf
import tflearn

X = [[0., 0.], [0., 1.], [1., 0.], [1., 1.]]
Y_xor = [[0.], [1.], [1.], [0.]]

# Graph definition
with tf.Graph().as_default():
    tnorm = tflearn.initializations.uniform(minval=-1.0, maxval=1.0)
    net = tflearn.input_data(shape=[None, 2], name='inputLayer')
    net = tflearn.fully_connected(net, 2, activation='sigmoid', weights_init=tnorm, name='layer1')
    net = tflearn.fully_connected(net, 1, activation='softmax', weights_init=tnorm, name='layer2')
    regressor = tflearn.regression(net, optimizer='sgd', learning_rate=2., loss='mean_square', name='layer3')

    # Training
    m = tflearn.DNN(regressor)
    m.fit(X, Y_xor, n_epoch=100, snapshot_epoch=False) 

    # Testing
    print("Testing XOR operator")
    print("0 xor 0:", m.predict([[0., 0.]]))
    print("0 xor 1:", m.predict([[0., 1.]]))
    print("1 xor 0:", m.predict([[1., 0.]]))
    print("1 xor 1:", m.predict([[1., 1.]]))
    
    layer1_var = tflearn.variables.get_layer_variables_by_name('layer1')
    layer2_var = tflearn.variables.get_layer_variables_by_name('layer2')
    inputLayer_var = tflearn.variables.get_layer_variables_by_name('inputLayer')
    
    #result = tf.matmul(inputLayer_var, layer1_var[0]) + layer1_var[1]
    
    with m.session.as_default():
    	print(tflearn.variables.get_value(layer1_var[0]))
    	print(tflearn.variables.get_value(layer1_var[1]))
    	print(tflearn.variables.get_value(layer2_var[0]))
    	print(tflearn.variables.get_value(layer2_var[1]))
```

**Comment by aymericdamien** [comment]
You just need to create a new model, you can do something like that:
```python
import tensorflow as tf
import tflearn

X = [[0., 0.], [0., 1.], [1., 0.], [1., 1.]]
Y_xor = [[0.], [1.], [1.], [0.]]

# Graph definition
with tf.Graph().as_default():
    tnorm = tflearn.initializations.uniform(minval=-1.0, maxval=1.0)
    net = tflearn.input_data(shape=[None, 2], name='inputLayer')
    layer1 = tflearn.fully_connected(net, 2, activation='sigmoid', weights_init=tnorm, name='layer1')
    layer2 = tflearn.fully_connected(layer1, 1, activation='softmax', weights_init=tnorm, name='layer2')
    regressor = tflearn.regression(layer2, optimizer='sgd', learning_rate=2., loss='mean_square', name='layer3')

    # Training
    m = tflearn.DNN(regressor)
    m.fit(X, Y_xor, n_epoch=100, snapshot_epoch=False) 

    # Testing
    print("Testing XOR operator")
    print("0 xor 0:", m.predict([[0., 0.]]))
    print("0 xor 1:", m.predict([[0., 1.]]))
    print("1 xor 0:", m.predict([[1., 0.]]))
    print("1 xor 1:", m.predict([[1., 1.]]))

    # You can create a new model, that share the same session (to get same weights)
    # Or you can also simply save and load a model
    m2 = tflearn.DNN(layer1, session=m.session)
    print(m2.predict([[0., 0.]]))
```

**Comment by krayush07** [comment]
Thanks. I am able to compute hidden state values as mentioned. However, I have one concern with this method. Doesn't it mean that if we need to have outputs from different layers, we need to re-compute (_layer_name.predict()_) the previous layers to reach current '_layer_name_' every time?

Isn't there a way around to do it in a single pass ('_final_layer.predict_') ?

**Comment by aymericdamien** [comment]
yes, you can, you just need to use a 'merge' layer, and use that layer as your 'regression' layer entry.

**Comment by krayush07** [comment]
Could you please elaborate it?

Suppose, I have a LSTM layer with a softmax layer for prediction. Now let's say I have a sentence sequence [21, 32, 11, 34, 56] as the test input. I need to have LSTM prediction for each case [21], [21, 32], [21, 32, 11], [21, 32, 11, 56] as test input. One way is to simply create such mapping exclusively in the test input itself and pass it one by one. However, since LSTM itself uses previous context to predict incoming label/word/token, so can we use the LSTM layer output by setting return_seq FLAG and feed each of the five tensor to the softmax layer individually without re-iterating as separate instances?

```
input_net = tflearn.input_data([None, 4, 3]) # 4 sentence size, 3 feature size
#lstm_net = tflearn.layers.recurrent.lstm(input_net, 100, dropout=0.8, return_seq=False, return_state=False, dynamic=True, name='lstm_layer')
lstm_net = tflearn.layers.recurrent.lstm(input_net, 10, dropout=0.8, name='lstm_layer', return_seq=True)
softmax_net = tflearn.fully_connected(lstm_net[3], 3, activation='softmax', name='softmax_layer')
regression_net = tflearn.regression(softmax_net, optimizer='adam', learning_rate=0.001, loss='categorical_crossentropy')

# Training
model = tflearn.DNN(regression_net, tensorboard_verbose=0)

with tf.Session() as sess:
	trainInput = sess.run([inputMatrix], feed_dict={embedding_placeholder: embedding, id_placeholder: trainX})
	print(trainInput)
	trainInput = np.reshape(trainInput, (-1, 4, 3))
	model.fit(trainInput, trainY)
	
	testX = np.genfromtxt('testX.csv', delimiter=',')
	test_arr = sess.run([inputMatrix], feed_dict={embedding_placeholder: embedding, id_placeholder: testX})
	print(model.predict(np.reshape(test_arr, (-1, 4, 3))))  #can we have softmax prediction for each of the subsequent increasing sequence?
```

**Comment by aymericdamien** [comment]
I am not exactly sure about what you want, is it only a way to retrieve the output of multiple layers with a single 'run'? To do this, from my example, you can write:
```python
net = ...
merge_op = tflearn.merge([softmax, layer1, layer2], axis=1)
regressor = tflearn.regression(softmax_net, optimizer='adam', learning_rate=0.001, loss='categorical_crossentropy')
m = tflearn.DNN(regressor)
```
And then in your output you will have all the selected layers computation.

**Comment by krayush07** [comment]
The answer you gave is about the case wherein I can have multiple layers output in single run. 

However I am asking for the case of LSTM wherein I need to have separate output for each successive timestamp for each of the instances in the test sequence. The comment above may provide an insight about what I want to achieve.

Suppose, I have a LSTM layer with a softmax layer for prediction. Now let's say I have a sentence sequence [21, 32, 11, 34, 56] as the test input. I want to have LSTM prediction for each case [21], [21, 32], [21, 32, 11], [21, 32, 11, 56] as test input. One way is to simply create such instances exclusively in the test input itself and pass it one by one. However, since LSTM itself uses previous context to predict incoming label/word/token, so can we use the LSTM layer output by setting return_seq FLAG and feed each of the five tensor to the softmax layer individually without re-iterating as separate instances.

**Comment by aymericdamien** [comment]
yes, I was editing my reply ^^ you can try this:
```python
net = lstm(net, 32, return_seq=True)
net = tf.pack(net) # You regroup everything into a single tensor
regressor = tflearn.regressor(net)
```

**Comment by krayush07** [comment]
I am afraid that won't solve my case, I guess.

What I want to say is that: if I have to predict next word/token for a sequence, say [1,21,13,4], [1,21,13,4,15], [1,21,13,4,52] for example, can we get a output from the common sequence of [1,21,13,4] and plug it into other two cases along-with input of [15] and [52] rather than re-iterating on the complete sequence as a whole.

**Comment by krayush07** [comment]
I am not able to achieve the thing. I meant that if I have a sequence say [1, 21, 34, 22], I wish to have softmax prediction for each of the LSTM unit. Currently I am able to extract the prediction from the final unit only.

**Comment by anirudhjoshi95** [comment]
@aymericdamien: I understand how you can do this by continuing the same session but how would you do it by saving and loading the model?
